tests/put_get.go

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2023 MinIO, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package tests
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/minio/confess/utils"
+	"github.com/minio/minio-go/v7"
+)
+
+const objectCountForPutGetTest = 50
+
+// PutGetTest uploads the objects using a random client and then
+// Gets the objects from each node to verify if the get output
+// is consistent accross nodes.
+type PutGetTest struct {
+	clients             []*minio.Client
+	logFile             *os.File
+	bucket              string
+	objectsWithMetaData sync.Map
+	apiStats            *Stats
+	concurrency         int
+}
+
+// metaData to store the uploaded object's info
+// which will be used for verification.
+type metaData struct {
+	sha256 []byte
+	size   int64
+	etag   string
+}
+
+func (t *PutGetTest) bucketName() string {
+	return t.bucket
+}
+
+// Name - name of the test.
+func (t *PutGetTest) Name() string {
+	return "PutGet"
+}
+
+// Init initialized the test.
+func (t *PutGetTest) Init(ctx context.Context, config Config, stats *Stats) error {
+	t.clients = config.Clients
+	t.logFile = config.LogFile
+	t.bucket = config.Bucket
+	t.objectsWithMetaData = sync.Map{}
+	t.apiStats = stats
+	t.concurrency = config.Concurrency
+	return nil
+}
+
+// Setup prepares the test by uploading the objects to the test bucket.
+func (t *PutGetTest) Setup(ctx context.Context) error {
+	return runFn(ctx,
+		objectCountForPutGetTest,
+		t.concurrency,
+		func(ctx context.Context, index int) error {
+			client, err := utils.GetRandomClient(t.clients)
+			if err != nil {
+				return err
+			}
+			object := fmt.Sprintf("confess/%s/%s/pfx-%d", t.Name(), time.Now().Format("01_02_06_15:04"), index)
+			hasher := sha256.New()
+			tr := io.TeeReader(reader(humanize.MiByte*5), hasher)
+			if objectInfo, err := put(ctx, putConfig{
+				client:     client,
+				bucketName: t.bucket,
+				objectName: object,
+				reader:     tr,
+				opts:       minio.PutObjectOptions{},
+			}, t.apiStats); err != nil {
+				if err := log(ctx,
+					t.logFile,
+					t.Name(),
+					client.EndpointURL().String(),
+					fmt.Sprintf("unable to put the object %s; %v", object, err),
+				); err != nil {
+					return err
```

Comment:
for errors in setup phase due to connection refused/node offline, you can just abandon the entire test - no need to log to file.

Reply:
> for errors in setup phase due to connection refused/node offline

Reply:
you could either use another online node/ abandon the test - our intent is to upload to one node and verify consistency across all nodes

Reply:
yes, currently we log that error and continue uploading the other objects using a random online client.

Reply:
there's no necessity to log setup errors - we want to capture only actual consistency issues, the tool should retry other failures if it occurs during the actual test/verify phase